### PR TITLE
add quantization transform

### PR DESCRIPTION
**Pull request by t-vi**
This adds a transform with 4 bit quantization transform using bits and bytes.

### CODE REVIEW

File: thunder/transforms/quantization.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+        checks = get_checks(prologue_trace)
+
+        compute_producers, compute_consumers = utils.producers_and_consumers(computation_trace)
+
+        if epilogue_trace is None:
+            prologue_to_epilogue_outputs = prologue_trace.output
+            output_subindex = None
+        else:
+            prologue_to_epilogue_outputs = prologue_trace.output[0]
```
Comment (lantiga) [inline thunder/transforms/quantization.py:103]: Just for my edification, can you explain these 4 lines?

File: thunder/transforms/quantization.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            new_computation_trace.bound_symbols.append(bsym.from_bsym())
+        new_computation_trace.bound_symbols += new_bindings
+        proxies_to_replace = {}
+        for bsym in computation_trace.bound_symbols[idx:]:
+            if bsym.sym == thunder.torch.linear and id(bsym.args[1]) in quantized_proxies:
+                assert len(bsym.args) == 3  # torch.linear(input, weight, bias)
+                n = quantized_proxies[id(bsym.args[1])]
+                qs = self.quant_states[n]
+                # bnb_matmul_nf4(x, qweight, bias, absmax, quant_map, blocksize, dtype, shape)
+                #  how to replace the _call_ctx(???)
```
Comment (lantiga) [inline thunder/transforms/quantization.py:176]: can you expand the comment?

**Review by lantiga — APPROVED** [review]
Looks awesome, super clean

File: thunder/transforms/quantization.py (hunk 1 above)
Comment (t-vi) [reply thunder/transforms/quantization.py:103]: Ha, glad you asked after I filed #628 :) This is for traces with / without epilogues.

Comment (t-vi) [reply thunder/transforms/quantization.py:105]: I added a comment.

File: thunder/tests/test_networks.py [hunk 3]
```diff
@@ -204,3 +204,62 @@ def test_hf_bart_self_attn():
...
+    model_fp_reference.eval()
+
+    jm = thunder.jit(
+        model_fp_reference,
+        executors=(bitsandbytes_executor,),
+        early_transforms=[BitsAndBytesLinearQuant4bit()],
+    )
+
+    logits_thunder = jm(x, input_pos)
+    # check_dtype=False due to litgpt returning float32
```
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/tests/test_networks.py:256]: I think I also got float32 when I did pass `precision=bf16-true` and no plugin, but that could have been an older version of lightning doing it.

**Pull request merged**